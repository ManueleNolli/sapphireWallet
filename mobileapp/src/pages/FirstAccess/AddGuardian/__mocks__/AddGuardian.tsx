const AddGuardianMock = () => <div data-testid={'AddGuardian'}>AddGuardian</div>

export default AddGuardianMock
