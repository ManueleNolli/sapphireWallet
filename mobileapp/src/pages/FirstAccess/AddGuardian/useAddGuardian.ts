import { useContext, useState } from 'react'
import { FirstAccessContext } from '../../../context/FirstAccessContext'
import { AddGuardianProps } from '../../../navigation/FirstAccessStack'
import useLoading from '../../../hooks/useLoading'
import { requestContractWallet } from '../../../services/wallet'
import { WalletContext } from '../../../context/WalletContext'
import { BlockchainContext } from '../../../context/BlockchainContext'
import { ZeroAddress } from 'ethers'
import Toast from 'react-native-toast-message'

export default function useAddGuardian(navigation: AddGuardianProps['navigation']) {
  const { toggleFirstAccess } = useContext(FirstAccessContext)
  const { currentNetwork } = useContext(BlockchainContext)
  const { getEOAAddress, setWalletContractAddress } = useContext(WalletContext)
  const { isLoading, setIsLoading } = useLoading(false)
  const [valueAddress, setValueAddress] = useState<string>('')
  const [isAddressValid, setIsAddressValid] = useState<boolean>(false)
  const [isQRCodeScanning, setIsQRCodeScanning] = useState<boolean>(false)

  const skipGuardian = () => {
    finishFirstAccess(ZeroAddress)
  }

  const withGuardian = () => {
    finishFirstAccess(valueAddress)
  }

  const finishFirstAccess = async (guardian: string) => {
    setIsLoading(true)
    try {
      const contractWalletAddress = await requestContractWallet(currentNetwork, getEOAAddress(), guardian)
      await setWalletContractAddress(contractWalletAddress.address)
      Toast.show({
        type: 'success',
        text1: 'Account created! ðŸ‘›',
      })
      await toggleFirstAccess()
    } catch (e: any) {
      Toast.show({
        type: 'error',
        text1: 'Transaction failed! ðŸ˜¢',
        text2: e.message,
      })
      setIsLoading(false)
      navigation.goBack()
    }
  }

  const QRCodeFinishedScanning = (data: string) => {
    setValueAddress(data)
    setIsQRCodeScanning(false)
  }

  const closeQRCodeScanner = () => {
    setIsQRCodeScanning(false)
  }

  return {
    valueAddress,
    setValueAddress,
    isAddressValid,
    setIsAddressValid,
    isQRCodeScanning,
    closeQRCodeScanner,
    setIsQRCodeScanning,
    QRCodeFinishedScanning,
    skipGuardian,
    withGuardian,
    isLoading,
  }
}
