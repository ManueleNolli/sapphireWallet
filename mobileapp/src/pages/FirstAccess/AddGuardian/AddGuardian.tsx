import React from 'react'

import { Button, Layout, Modal, Text } from '@ui-kitten/components'
import SafeAreaView from '../../../utils/SafeAreaView'
import { StyleSheet, View } from 'react-native'
import { vh, vw } from '../../../Styles'
import { AddGuardianProps } from '../../../navigation/FirstAccessStack'
import Loading from '../../Loading/Loading'
import useAddGuardian from './useAddGuardian'
import InputAddress from '../../../components/InputAddress/InputAddress'
import QRCodeScanner from '../../../components/QRCodeScanner/QRCodeScanner'

export default function AddGuardian({ navigation }: AddGuardianProps) {
  const {
    valueAddress,
    setValueAddress,
    isAddressValid,
    setIsAddressValid,
    isQRCodeScanning,
    setIsQRCodeScanning,
    closeQRCodeScanner,
    QRCodeFinishedScanning,
    withGuardian,
    skipGuardian,
    isLoading,
  } = useAddGuardian(navigation)

  if (isLoading) {
    return <Loading text="Deploying smart contract wallet..." />
  }

  if (isQRCodeScanning)
    return (
      <Modal
        visible={isQRCodeScanning}
        backdropStyle={{ backgroundColor: 'rgba(0, 0, 0, 0.5)' }}
        onBackdropPress={closeQRCodeScanner}
      >
        <QRCodeScanner onQRCodeScanned={QRCodeFinishedScanning} />
      </Modal>
    )

  return (
    <Layout style={{ flex: 1 }}>
      <SafeAreaView style={{ flex: 1, marginVertical: 20 * vh }}>
        <View
          style={{
            justifyContent: 'center',
            alignItems: 'center',
            marginVertical: 2 * vh,
          }}
        >
          <Text category="h4" style={styles.text}>
            Add a guardian
          </Text>
          <Text category="p1" style={[styles.text, styles.descriptionText]}>
            Guardians are able to recover your wallet if you lose your phone. You can add more guardians later.
          </Text>
        </View>
        <InputAddress
          label=""
          value={valueAddress}
          setValue={setValueAddress}
          isValid={isAddressValid}
          setIsValid={setIsAddressValid}
          setIsQRCodeScanning={setIsQRCodeScanning}
        />
        <View>
          <Button appearance="outline" onPress={withGuardian} style={{ marginVertical: 2 * vh }}>
            Create wallet
          </Button>

          <Button appearance="ghost" onPress={skipGuardian} style={{ width: 25 * vw, alignSelf: 'flex-end' }}>
            Skip
          </Button>
        </View>
      </SafeAreaView>
    </Layout>
  )
}

const styles = StyleSheet.create({
  text: {
    textAlign: 'justify',
  },
  descriptionText: {
    textAlign: 'center',
    marginVertical: 2 * vh,
  },
})
