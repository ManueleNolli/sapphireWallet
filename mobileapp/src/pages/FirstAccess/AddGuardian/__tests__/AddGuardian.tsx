import renderWithTheme from '../../../../TestHelper'
import useAddGuardian from '../useAddGuardian'
import AddGuardian from '../AddGuardian'

jest.mock('../useAddGuardian', () => jest.fn())
jest.mock('../../../../components/QRCodeScanner/QRCodeScanner')

describe('AddGuardian', () => {
  it('renders correctly', () => {
    ;(useAddGuardian as jest.Mock).mockReturnValue({
      valueAddress: '',
      setValueAddress: jest.fn(),
      isAddressValid: true,
      setIsAddressValid: jest.fn(),
      isQRCodeScanning: false,
      setIsQRCodeScanning: jest.fn(),
      QRCodeFinishedScanning: jest.fn(),
      skipGuardian: jest.fn(),
      withGuardian: jest.fn(),
      isLoading: false,
    })

    const tree = renderWithTheme(
      <AddGuardian
        {...({
          route: {},
        } as any)}
      />
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when loading', () => {
    ;(useAddGuardian as jest.Mock).mockReturnValue({
      valueAddress: '',
      setValueAddress: jest.fn(),
      isAddressValid: true,
      setIsAddressValid: jest.fn(),
      isQRCodeScanning: false,
      setIsQRCodeScanning: jest.fn(),
      QRCodeFinishedScanning: jest.fn(),
      skipGuardian: jest.fn(),
      withGuardian: jest.fn(),
      isLoading: true,
    })

    const tree = renderWithTheme(
      <AddGuardian
        {...({
          route: {},
        } as any)}
      />
    )

    expect(tree).toMatchSnapshot()
    expect(tree.getByText('Deploying smart contract wallet...')).toBeTruthy()
  })

  it('renders correctly when scanning', () => {
    ;(useAddGuardian as jest.Mock).mockReturnValue({
      valueAddress: '',
      setValueAddress: jest.fn(),
      isAddressValid: true,
      setIsAddressValid: jest.fn(),
      isQRCodeScanning: true,
      setIsQRCodeScanning: jest.fn(),
      QRCodeFinishedScanning: jest.fn(),
      skipGuardian: jest.fn(),
      withGuardian: jest.fn(),
      isLoading: false,
    })

    const tree = renderWithTheme(
      <AddGuardian
        {...({
          route: {},
        } as any)}
      />
    )

    expect(tree).toMatchSnapshot()
  })
})
