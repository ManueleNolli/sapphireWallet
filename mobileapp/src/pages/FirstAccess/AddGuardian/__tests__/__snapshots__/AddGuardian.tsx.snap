// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddGuardian renders correctly 1`] = `
<View
  appearance="default"
  level="1"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    appearance="default"
    level="1"
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "flex": 1,
            "marginHorizontal": 37.5,
            "marginVertical": 66.7,
          },
          {
            "flex": 1,
            "marginVertical": 266.8,
          },
        ]
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
            "marginVertical": 26.68,
          }
        }
      >
        <Text
          appearance="default"
          category="h4"
          status="basic"
          style={
            [
              {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 26,
                "fontWeight": "800",
              },
              {
                "textAlign": "justify",
              },
            ]
          }
        >
          Add a guardian
        </Text>
        <Text
          appearance="default"
          category="p1"
          status="basic"
          style={
            [
              {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              [
                {
                  "textAlign": "justify",
                },
                {
                  "marginVertical": 26.68,
                  "textAlign": "center",
                },
              ],
            ]
          }
        >
          Guardians are able to recover your wallet if you lose your phone. You can add more guardians later.
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="@undefined/container"
      >
        <View
          style={
            [
              {
                "backgroundColor": "#F7F9FC",
                "borderColor": "#00E096",
                "borderRadius": 4,
                "borderWidth": 1,
                "minHeight": 40,
                "paddingHorizontal": 8,
                "paddingVertical": 7,
              },
              {
                "alignItems": "center",
                "flexDirection": "row",
                "width": "100%",
              },
            ]
          }
        >
          <TextInput
            appearance="default"
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            placeholder="Address"
            placeholderTextColor="#8F9BB3"
            size="medium"
            status="success"
            style={
              [
                {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "normal",
                  "marginHorizontal": 8,
                },
                {
                  "flexBasis": "auto",
                  "flexGrow": 1,
                  "flexShrink": 1,
                },
                null,
                undefined,
              ]
            }
            testID="@undefined/input"
            value=""
          />
          <Image
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            source={1}
            style={
              {
                "height": 24,
                "marginHorizontal": 8,
                "tintColor": "#00E096",
                "width": 24,
              }
            }
            testID="qr-code-icon"
          />
        </View>
      </View>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(51, 102, 255, 0.08)",
              "borderColor": "#3366FF",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginVertical": 26.68,
              "minHeight": 40,
              "minWidth": 40,
              "opacity": 1,
              "paddingHorizontal": 10,
              "paddingVertical": 12,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            status="basic"
            style={
              [
                {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                {
                  "color": "#3366FF",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "marginHorizontal": 10,
                },
              ]
            }
          >
            Create wallet
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "flex-end",
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 40,
              "minWidth": 40,
              "opacity": 1,
              "paddingHorizontal": 10,
              "paddingVertical": 12,
              "width": 187.5,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            status="basic"
            style={
              [
                {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                {
                  "color": "#3366FF",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "marginHorizontal": 10,
                },
              ]
            }
          >
            Skip
          </Text>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;

exports[`AddGuardian renders correctly when loading 1`] = `
<View
  appearance="default"
  level="1"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <ViewManagerAdapter_ExpoImage
    contentFit="cover"
    contentPosition={
      {
        "left": "50%",
        "top": "50%",
      }
    }
    height={375}
    maxHeight={250}
    maxWidth={250}
    onError={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    placeholder={[]}
    source={[]}
    style={
      {
        "height": 375,
        "maxHeight": 250,
        "maxWidth": 250,
        "width": 375,
      }
    }
    transition={null}
    width={375}
  />
  <Text
    appearance="default"
    category="h6"
    status="basic"
    style={
      [
        {
          "color": "#222B45",
          "fontFamily": "System",
          "fontSize": 18,
          "fontWeight": "800",
        },
        undefined,
      ]
    }
  >
    Deploying smart contract wallet...
  </Text>
</View>
`;

exports[`AddGuardian renders correctly when scanning 1`] = `
<Modal
  hardwareAccelerated={false}
  statusBarTranslucent={false}
  transparent={true}
  visible={true}
>
  <View
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "color-basic-transparent-600",
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="@backdrop"
    />
    <View
      backdropStyle={
        {
          "backgroundColor": "color-basic-transparent-600",
        }
      }
      onLayout={[Function]}
      shouldUseContainer={true}
      style={
        [
          undefined,
          {
            "position": "absolute",
          },
          {
            "left": -999,
            "top": -999,
          },
        ]
      }
      visible={true}
    >
      <div
        data-testid="QRCodeScanner"
      >
        QRCodeScanner
      </div>
    </View>
  </View>
</Modal>
`;
